Examples/ComplexVaporExample/Sources/App/routes.swift

```diff
@@ -0,0 +1,114 @@
+import MongoSwift
+import Vapor
+
+/// Constructs a document using the ID from the specified request which can be used a filter
+/// for MongoDB reads/updates/deletions.
+func getIDFilter(from request: Request) throws -> BSONDocument {
+    // We only call this method from request handlers that have _id parameters so the value
+    // will always be available.
+    guard let idString = request.parameters.get("_id") else {
+        throw Abort(.internalServerError, reason: "Request unexpectedly missing _id parameter")
+    }
+    // If the BSONObjectID constructor throws, the ID is invalid.
+    guard let id = try? BSONObjectID(idString) else {
+        throw Abort(.notFound, reason: "Invalid ID \(idString)")
+    }
+    return ["_id": .objectID(id)]
+}
+
+/// Adds a collection of routes to the application.
+func routes(_ app: Application) throws {
+    /// A collection with type `Kitten`. This allows us to directly retrieve instances of
+    /// `Kitten` from the collection.  `MongoCollection` is safe to share across threads.
+    let collection = app.mongoClient.db("home").collection("kittens", withType: Kitten.self)
+
+    /// Handles a request to load the main index page containing a list of kittens.
+    app.get { req -> EventLoopFuture<View> in
+        collection.find().flatMap { cursor in
+            cursor.toArray()
+        // Hop to ensure that the final response future happens on the request's event loop.
+        }.hop(to: req.eventLoop)
+        .flatMap { kittens in
+            // Return the corresponding Leaf view, providing the list of kittens as context.
+            req.view.render("index.leaf", IndexContext(kittens: kittens))
+        }
+        .flatMapErrorThrowing { error in
+            throw Abort(.internalServerError, reason: "Failed to load kittens: \(error)")
+        }
+    }
+
+    /// Handles a request to add a new kitten.
+    app.post { req -> EventLoopFuture<Response> in
+        let newKitten = try req.content.decode(Kitten.self)
+        return collection.insertOne(newKitten)
+        // Hop to ensure that the final response future happens on the request's event loop.
+        .hop(to: req.eventLoop)
+        .map { _ in
+            // On success, redirect to the index to reload the updated list.
+            req.redirect(to: "/")
+        }
+        .flatMapErrorThrowing { error in
+            // Give a more helpful error message in case of a duplicate key error.
+            if let err = error as? MongoSwift.WriteError, err.writeFailure?.code == 11000 {
```

Comment:
nit: `MongoSwift.MongoError.WriteError`

Reply:
good catch, that wasn't merged yet when I started working on this and I hadn't pulled in the new changes!